### PR TITLE
(#857) Load Chocolatey executable assembly

### LINKED ISSUE #857: Load Chocolatey main assembly during runtime
When running Chocolatey GUI and Chocolatey GUI CLI we should load the main assembly (`choco.exe`) instead of running the statically referenced DLL file we have compiled against.

This will allow changes to Chocolatey extensions to make use of new functionality in Chocolatey, or include any bug fixes in Chocolatey without having to publish a new version of Chocolatey GUI.

### PR DESCRIPTION
**Pull request by AdmiringWorm**
This pull request updates the Chocolatey GUI and Chocolatey GUI CLI Assembly Loader to load the main Chocolatey executable as an assembly, instead of loading the included Chocolatey DLL file.
With the changes made here, the DLL file will no longer be included in the installation.

fixes #857

### CODE REVIEW

**Comment by gep13** [comment]
@AdmiringWorm this change will go out as a hotfix release, rather than as part of the next release from `develop` branch.  Can you create a new `hotfix/0.18.2` branch, and re-target this PR there?

File: Source/ChocolateyGui/App.xaml.cs [hunk 1]
```diff
@@ -32,30 +32,49 @@ public App()
...
+                            installDir = Path.Combine(rootDrive, "ProgramData", "chocolatey");
+                        }
+
+                        var assemblyLocation = Path.Combine(installDir, "choco.exe");
+
+                        return AssemblyResolver.ResolveOrLoadAssembly("choco", string.Empty, assemblyLocation);
+                    }
+
 #if FORCE_CHOCOLATEY_OFFICIAL_KEY
                 var chocolateyGuiPublicKey = Bootstrapper.OfficialChocolateyPublicKey;
```
Comment (gep13) [inline Source/ChocolateyGui/App.xaml.cs:57]: Small nit pick, but can we bump this var over a bit as well?

File: Source/ChocolateyGui/App.xaml.cs [hunk 2]
```diff
@@ -32,30 +32,49 @@ public App()
...
 #else
-                var chocolateyGuiPublicKey = Bootstrapper.UnofficialChocolateyPublicKey;
+                    var chocolateyGuiPublicKey = Bootstrapper.UnofficialChocolateyPublicKey;
 #endif
 
-                try
-                {
-                    if (requestedAssembly.get_public_key_token().is_equal_to(chocolateyGuiPublicKey)
-                        && requestedAssembly.Name.is_equal_to(Bootstrapper.ChocolateyGuiCommonAssemblySimpleName))
+                    if (AssemblyResolver.IsPublicKeyToken(requestedAssembly, chocolateyGuiPublicKey)
```
Comment (gep13) [inline Source/ChocolateyGui/App.xaml.cs:62]: `IsPublicKeyToken` not sure about the name here. You are comparing one public key token to another, right? Should this be something like `DoesPublicKeyTokenMatch`? Naming is hard 😢

File: Source/ChocolateyGui.Common/Startup/AssemblyResolver.cs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            return ResolveOrLoadAssemblyInternal(
+                "chocolatey",
+                string.Empty,
+                assemblyFileLocation,
+                (assembly) =>
+                {
+                    var assemblyName = assembly.GetName();
+
+                    return string.Equals(assemblyName.Name, "chocolatey", StringComparison.OrdinalIgnoreCase) &&
+                        assemblyName.Version != new Version(0, 10, 15, 0); // TODO: Should maybe be created automatically
```
Comment (gep13) [inline Source/ChocolateyGui.Common/Startup/AssemblyResolver.cs:30]: Not sure what you mean here by this TODO. Can you elaborate?

File: Source/ChocolateyGuiCli/Runner.cs [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+                }
+
+                var configuration = new ChocolateyGuiConfiguration();
+                SetUpGlobalOptions(args, configuration, Bootstrapper.Container);
+                SetEnvironmentOptions(configuration);
+
+                if (configuration.RegularOutput)
+                {
+#if DEBUG
+                Bootstrapper.Logger.Warning(" (DEBUG BUILD)".format_with("Chocolatey GUI", configuration.Information.DisplayVersion));
```
Comment (gep13) [inline Source/ChocolateyGuiCli/Runner.cs:66]: Can we bump this over a bit?

**Review by gep13 — CHANGES_REQUESTED** [review]
This looks really good to me, just a couple of minor tweaks and suggestions.  Can you update the commit message with information about _why_ we need this change?  i.e. to protect ourselves from incompatibilities when the version of Chocolatey installed globally is newer than what is shipped with Chocolatey GUI.

**Comment by gep13** [comment]
@AdmiringWorm am I right in saying that the changes that are included here result in the chocolatey.dll being omitted from the generated MSI, or is there additional work needed here?

File: Source/ChocolateyGui.Common/Startup/AssemblyResolver.cs (hunk 3 above)
Comment (AdmiringWorm) [reply Source/ChocolateyGui.Common/Startup/AssemblyResolver.cs:30]: The initial intention was that it should only try to load the `choco.exe` if it was not the version we already referenced, however, this is leftover from the initial testing and I need to remove this function completely.

**Comment by AdmiringWorm** [comment]
> @AdmiringWorm am I right in saying that the changes that are included here result in the chocolatey.dll being omitted from the generated MSI, or is there additional work needed here?

There is no additional work needed, due to all of the projects setting `chocolatey.dll` to not copy itself it won't end up in the MSI either.

Requested changes have been made.

**Comment by AdmiringWorm** [comment]
> @AdmiringWorm this change will go out as a hotfix release, rather than as part of the next release from develop branch. Can you create a new hotfix/0.18.2 branch, and re-target this PR there?

Completely missed this comment initially, I can unfortunately not create a hotfix branch due to lacking permissions.

**Review by gep13 — APPROVED** [review]
LGTM!

**Comment by gep13** [comment]
@AdmiringWorm thank you very much for working through this!

**Pull request merged**